File: day20.py
from collections import deque, defaultdict
from math import lcm

modules = {}
feeder = ""
while line := input():
    name, adj = line.split(" -> ")
    adj = adj.split(", ")

    if name == "broadcaster":
        modules[name] = ("", adj)
    else:
        module_type = name[0]
        name = name[1:]
        modules[name] = (module_type, adj)
    
    if "rx" in adj:
        feeder = name

waiting = set()
ff_state = defaultdict(bool)
c_state = defaultdict(set)

for name, data in modules.items():
    for next in data[1]:
        if next == "rx": continue
        if modules[next][0] == "&":
            c_state[name].add(name)
        if next == feeder:
            waiting.add(name)

low, high, t, cycle_lcm = 0, 0, 0, 1
while True:
    t += 1
    bfs = deque([("broadcaster", "button", False)])
    while bfs:
        name, parent, pulse = bfs.popleft()
        if pulse: high += 1
        else: low += 1

        if name == "rx": continue
        module_type, adj = modules[name]
        if module_type == '%':
            if pulse: continue
            ff_state[name] = not ff_state[name]
            pulse = ff_state[name]
        elif module_type == '&':
            if pulse:
                c_state[name].discard(parent)
            else:
                c_state[name].add(parent)
            pulse = len(c_state[name]) > 0

        if pulse and name in waiting:
            waiting.remove(name)
            cycle_lcm = lcm(cycle_lcm, t)
            if not waiting:
                print(cycle_lcm)
                exit(0)

        for next_item in adj:
            bfs.append((next_item, name, pulse))

    if t == 1000:
        print(low * high)

File: day20.cpp
#include <bits/stdc++.h>
using namespace std;
#define INPUT(var) for(string var; getline(cin, var), var!="";)
#define FOR(i, s, e) for (int i=s; i<e; ++i)

int main() {
	map<string, pair<char, vector<string>>> modules;
	string feeder;
	INPUT(line) {
		stringstream ss(line);
		string name; ss >> name;
		string next; ss >> next;
		if (name != "broadcaster") {
			char type = name[0];
			name.erase(name.begin());
			modules[name].first = type;
		}
		while (ss >> next) {
			if (next.back() == ',')
				next.pop_back();
			modules[name].second.push_back(next);
			if (next == "rx")
				feeder = name;
		}
	}

	
	set<string> waiting;
	map<string, bool> ff_state; // flip flop
	map<string, set<string>> c_state; // conjunction
	for (auto [name, _data] : modules) {
		for (string next : _data.second) {
			if (modules[next].first == '&')
				c_state[next].insert(name);
			if (next == feeder)
				waiting.insert(name);
		}
	}

	int low = 0, high = 0;
	long long cycle_lcm = 1;
	for (int t = 1;; ++t) {
		queue<tuple<string, string, bool>> bfs;
		bfs.push({ "broadcaster", "button", 0 });
		while (!bfs.empty()) {
			auto [name, parent, pulse] = bfs.front(); bfs.pop();
			pulse ? ++high : ++low;

			auto [type, adj] = modules[name];
			if (type == '%') {
				if (pulse) continue;
				pulse = ff_state[name] = !ff_state[name];
			}
			else if (type == '&') {
				if (pulse) c_state[name].erase(parent);
				else c_state[name].insert(parent);
				pulse = !c_state[name].empty();
			}

			if (pulse && waiting.count(name)) {
				waiting.erase(name);
				cycle_lcm = lcm(cycle_lcm, t);
				if (waiting.empty()) {
					cout << cycle_lcm << '\n';
					exit(0);
				}
			}

			for (string next : adj)
				bfs.push({ next, name, pulse });
		}

		if (t == 1000)
			cout << (long long)low * high << '\n';
	}
}